pkg/controller/deployment/deployment_helper.go

```diff
@@ -76,6 +76,74 @@ func parseArgMap(argMap map[string]string, args []string) {
 	}
 }
 
+// mergeContainerResources merges source container resources with that
+// provided as override resources.
+func mergeContainerResources(sourceResources corev1.ResourceRequirements, overrideResources corev1.ResourceRequirements) corev1.ResourceRequirements {
+	mergedResourceLimits := mergeContainerResourceList(sourceResources.Limits, overrideResources.Limits)
+	mergedResourceRequests := mergeContainerResourceList(sourceResources.Requests, overrideResources.Requests)
+	mergedResourceClaims := mergeResourceClaims(sourceResources.Claims, overrideResources.Claims)
+
+	return corev1.ResourceRequirements{
+		Limits:   mergedResourceLimits,
+		Requests: mergedResourceRequests,
+		Claims:   mergedResourceClaims,
+	}
+}
+
+// mergeContainerResourceList merges source resource list with that
+// provided as override resource list.
+func mergeContainerResourceList(sourceResourceList corev1.ResourceList, overrideResourceList corev1.ResourceList) corev1.ResourceList {
```

Comment:
Technically not a merge in case of `ResourceList` since we are basically overriding only here. So might as well rename this.

Reply:
Won't we merge the overriding resources with the source's resources if, say, only limits is provided in the overriding resources and not requests, whereas source's resources has both limits and requests?

If we override then whatever is in the overriding resources, even if either of empty limits or requests, will override the source's limits and requests even if something is already set there.

Reply:
Overriding resources takes priority so we set whatever is set there as the final value. What you are suggesting would be the case if one could update source's resources via a different channel as compared `overrideResources`.

Reply:
We will never have default resources, so yes. And if `overrideResources` is `nil` nothing to do but if `overrideResources` is `{}` then yes that will take precedence.

Reply:
I had a discussion with @TrilokGeer that we might want to do some study and set some default values in the future. For reference IBM's cert-manager operator sets some default values: https://github.com/IBM/ibm-cert-manager-operator/blob/master/controllers/resources/certmanager.go

Reply:
For requests we could have something in the future but for limits I'd still be hesitant to push during upgrades especially when majority of the users are using `Automatic` upgrades. This can be viewed as a potential breaking change during upgrades since we will not definitively know at what scale users are using cert-manager.  Limits are usually [frowned upon ](https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#resources-and-limits)on control plane components and although cert-manager doesn't affect control plane now it could in the future.

Reply:
Now that there is a possibility of adding default `requests` at least, I think we should be okay doing a `nil`/`empty` check before overriding so that we don't override with a nil/empty value for mem/cpu.

Reply:
> Limits are usually [frowned upon ](https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#resources-and-limits)on control plane components and although cert-manager doesn't affect control plane now it could in the future.

In the future, it is possible that API server endpoints use a cert-manager issued certificate and in that case if the operator/cert-manager operand goes down it could be a SLO problem.  I'd suggest preferring to refrain from adding any resource limits as of now (unless user themselves configures as such: which with this implementation they can do through overrideResources for operand and through Subscription object for the operator), also partly because upstream cert-manager also doesn't enforce these limits yet so we don't know what are good enough sane values for clusters with large certificate usage.

Reply:
So, if we set default requests and not limits, however, overriding resources has limits and no requests, do we remove the default requests and use whatever is in the overriding resources? Or do we keep the default resources and use the overriding limits?

Reply:
> In the future, it is possible that API server endpoints use a cert-manager issued certificate and in that case if the operator/cert-manager operand goes down it could be a SLO problem. I'd suggest preferring to refrain from adding any resource limits as of now (unless user themselves configures as such: which with this implementation they can do through overrideResources for operand and through Subscription object for the operator), also partly because upstream cert-manager also doesn't enforce these limits yet so we don't know what are good enough sane values for clusters with large certificate usage.

We might not set it right now, but the implementation should take into account the changes we might have in the future. So if we decide to add some default requests, an empty `overrideResources` will always end up removing that if we decide that it will override and not merge with default resources.

Reply:
Agreed, implementation should be in a way that it takes care of such scenarios.

I would say,
```yaml
originalDeployment:
  resources:
      requests: X
      limits: nil

overrides:
   resources:
     resources: nil
     limits: Y
```

should translate to:
```yaml
finalDeployment:
   resources:
     requests: X
     limits: Y
```
Hope it makes sense.

Reply:
Whereas,
```yaml
originalDeployment:
  resources:
      requests: X
      limits: nil

overrides:
   resources:
     resources: Z
     limits: Y
```

should still translate to:
```yaml
finalDeployment:
   resources:
     requests: Z
     limits: Y
```

Reply:
I was meaning to say, we stick to:
```yaml
finalDeployment:
  resources:
      requests:
        cpu: X
        memory: Y
      limits: nil
```
in such scenarios (which involves merging individual resource types, `iff not nil`).